### ISSUE TITLE
Renaming replaces references to unrelated but similarly named undeclared variables

### ISSUE THREAD

**Issue by schuelermine** [clangd/clangd#2097]
In the following erroneous code:

```c
void foo(void) {
    size_t some_a_value;
    size_t some_c_value = some_a_value + some_b_value;
}
```

…renaming `some_a_value` to `some_value` also replaces `some_b_value` with `some_value`.

[Screencast from 2024-07-18 23-36-44.webm](https://github.com/user-attachments/assets/a9aa44d0-ef8d-4778-b903-1a272428f116)

**System information**

Output of `clangd --version`: 17.0.6

Editor/LSP plugin: VSCode, v0.1.24

Operating system: NixOS unstable

**Comment by schuelermine** [comment]
A few more demonstrations

[Screencast from 2024-07-18 23-48-44.webm](https://github.com/user-attachments/assets/429b4762-197c-41bf-8cc8-174318128a09)
[Screencast from 2024-07-18 23-50-30.webm](https://github.com/user-attachments/assets/6da2d739-9dad-424f-aa35-7abc0eb9ac0f)
[Screencast from 2024-07-18 23-53-48.webm](https://github.com/user-attachments/assets/e761e7d6-a5b3-4af8-9a04-34571614ff98)

**Comment by HighCommander4** [comment]
This is caused by the clang parser's typo correction feature. Since `some_b_value` cannot be resolved, and `some_a_value` is close in edit distance, it gets typo-corrected `some_a_value`, and thus shows up as a reference to `some_a_value`.

Typo correction can be disabled by creating a `.clangd` file with:

```yaml
CompileFlags:
  Add: [-fno-spell-checking]
```

**Comment by schuelermine** [comment]
Thanks

**Comment by schuelermine** [comment]
I believe this is a potential source for bugs.

I refactored some code using mutexes to use one mutex instead of two. To do this, I deleted one mutex, and renamed the other. This caused clangd to rename all references to either mutex. I almost didn’t notice this, and it would’ve led to a double initialization and double destruction of these mutexes.

**Comment by HighCommander4** [comment]
I agree. As a user, I think what I'd like to see is to have clangd categorize the typo-corrected reference as a "tentative" reference (as opposed to one it's confident in), and prompt the user to review the tentative references before renaming them.

I'm not sure what abilities the Language Server Protocol gives a language server in this regard, but it's worth investigating. Please feel free to reopen the issue.

**Comment by schuelermine** [comment]
I believe this is a big enough problem that a good first step would be to inject `-fno-spell-checking` by default to mitigate this issue and consider alternatives thereafter.

**Comment by HighCommander4** [comment]
I don't think we'd want to disable typo correction altogether by default. That has impacts on other features than rename, in particular diagnostics -- in the event of an actual typo, typo correction can prevent a flood of cascading diagnostics that create noise.

A solution that's less involved than what I described above, but still scoped to the rename feature, could be to detect typo-corrected references and exclude them from being renamed for now.
